File: backend/app/api/routes/subjects.py
from typing import Annotated

from fastapi import APIRouter, Depends, status

from app.api.dependencies import get_subject_service
from app.schemas.common import MessageResponse
from app.schemas.subject import SubjectCreate, SubjectRead, SubjectUpdate
from app.services.subject import SubjectService

router = APIRouter(prefix="/subjects", tags=["subjects"])


@router.get("", response_model=list[SubjectRead])
async def list_subjects(
    service: Annotated[SubjectService, Depends(get_subject_service)],
) -> list[SubjectRead]:
    subjects = await service.list_subjects()
    return [SubjectRead.model_validate(subject) for subject in subjects]


@router.post("", response_model=SubjectRead, status_code=status.HTTP_201_CREATED)
async def create_subject(
    payload: SubjectCreate,
    service: Annotated[SubjectService, Depends(get_subject_service)],
) -> SubjectRead:
    subject = await service.create_subject(payload)
    return SubjectRead.model_validate(subject)


@router.put("/{subject_id}", response_model=SubjectRead)
async def update_subject(
    subject_id: int,
    payload: SubjectUpdate,
    service: Annotated[SubjectService, Depends(get_subject_service)],
) -> SubjectRead:
    subject = await service.update_subject(subject_id, payload)
    return SubjectRead.model_validate(subject)


@router.delete("/{subject_id}", response_model=MessageResponse)
async def delete_subject(
    subject_id: int,
    service: Annotated[SubjectService, Depends(get_subject_service)],
) -> MessageResponse:
    await service.delete_subject(subject_id)
    return MessageResponse(detail="Subject deleted successfully.")

File: frontend/src/api/subjects.ts
import http from '@/api/http'
import type { MessageResponse, Subject } from '@/types/models'

export const subjectsApi = {
  async list() {
    const { data } = await http.get<Subject[]>('/subjects')
    return data
  },
  async create(payload: Omit<Subject, 'id'>) {
    const { data } = await http.post<Subject>('/subjects', payload)
    return data
  },
  async update(id: number, payload: Partial<Omit<Subject, 'id'>>) {
    const { data } = await http.put<Subject>(`/subjects/${id}`, payload)
    return data
  },
  async remove(id: number) {
    const { data } = await http.delete<MessageResponse>(`/subjects/${id}`)
    return data
  },
}
